Confidential-Storage-Arch.md

```diff
@@ -0,0 +1,124 @@
+
+### Sources of Untrusted Storage Data
+
+Let's first enumerate the possible sources of untrusted storage data as below:
+- initrd (measured)
+- virtio-blk devices
+- virtio-scsi devices
+- virtio-pmem devices  
+- nvme devices
+- virtio-fs devices
+- virtio-9p devices
+- network attached filesystems, such as NFS, CephFS etc. There's a special case for network attached filesystems, it may
+  be trusted filesystem if it's connected to a trusted server through secure connections.
+- software services providing storage data, for example an image cache service integrated into the image service.  
+
+The storage subsystem may be composed up with multiple layers, so we may enforce confidentiality at different positions.
+Here confidentiality enforcement means decryption and/or integrity verification.
+
+### Enforce Confidentiality at The Data Source Layer
+
+The confidentiality may be enforced at the data source layer, such as at the service agents or the NFS clients.
+Possible storage stacks as below:
+
+![Enforce Confidentiality at Data Source Layer](images/storage-enforce-confidentiality-at-source.png)
+
+### Enforce Confidentiality at The Block Device Layer
+
+The confidentiality may also be enforced at the block device layer, such as by using the dm-crypt/dm-verity.
+Possible storage stacks as below:
+
+![Enforce Confidentiality at Block Device Layer](images/storage-enforce-confidentiality-at-blockdev.png)
+
+### Enforce Confidentiality at The FileSystem Layer
+
+The confidentiality may also be enforced at the filesystem layer, such as by using the to-be-invented filesystems.
+Possible storage stacks as below:
+
+![Enforce Confidentiality at Filesystem Layer](images/storage-enforce-confidentiality-at-fs.png)
+
+### Composition of Confidential Filesystem
+
+Once those confidential filesystems are ready, they could be composed up as normal filesystems in anyway, such as by
+mount, bind mount, overlayfs, chroot or pivot etc.
+
+## Ownership and Life Management of Storage Devices
+
+With traditional Kata Containers, it aims to be compatible with runC and all filesystems are prepared on host, 
+the Kata runtime just pass those storage resources into the guest. With confidential containers, that model doesn't
+work anymore, it needs to prepare filesystems for containers within the vm.
+
+In addition to normal container associated volume, a new concept of "pod built-in ephemeral volume/storage device" 
+is introduced. The pod built-in ephemeral volume  which will be used to:
+- host guest os temporary data 
+- provide ephemeral volumes for containers
+- store downloaded container images
+- store rw layer of container rootfs
+
+The lifecycle of the pod built-in ephemeral volume will be managed by the orchestration system or by the host.
+It may be associated with the pod/container, and get destroyed when deleting the pod/container. Or it may be reused
+for different pods/containers to improve performance.
+
+### Overall Architecture
+
```

Comment:
A more general comment here: We should also add a `Supported use cases` section and highlight 2-3 typical but concrete use cases that we want to initially support, describing e.g. how we would handle e.g. adding volumes to a confidential container/pod.

That would give us focus for the initial implementation steps. For example we could say that  the 3 initial scenarios that we want to support are:

1. Block device based volumes
2. Network attached volumes
3. Ephemeral volume

Going down into details for those 3 use cases would help us understand how things could be implemented.

I can work on that if we agree that would make sense.